### PR TITLE
na_position doesn't work for sort_index() with MultiIndex

### LINKED ISSUE #14784: na_position doesn't work for sort_index() with MultiIndex
#### Code Sample, a copy-pastable example if possible

```python
In [1]: import numpy as np

In [2]: import pandas as pd

In [3]: pd.__version__
Out[3]: u'0.19.0'

In [4]: mi = pd.MultiIndex.from_tuples([[1, 1, 3], [1, 1, np.nan]], names=list('ABC'))

In [5]: df = pd.DataFrame([[1, 2], [3, 4]], mi)

In [6]: df.sort_index(na_position="first")
Out[6]: 
         0  1
A B C        
1 1 NaN  3  4
    3    1  2

In [7]: df.sort_index(na_position="last")
Out[7]: 
         0  1
A B C        
1 1 NaN  3  4
    3    1  2
```

#### Problem description

The `na_position` argument isn't used in `DataFrame.sort_index()` or `Series.sort_index()` due to the way we sort the `MultiIndex`.  Whenever we create a `MultiIndex`, we store the labels as relative values.  For instance, if we have the following `MultiIndex`:  
```python
MultiIndex.from_tuples([[1, 1, 3], [1, 1, np.nan]], names=list('ABC'))
```
the values get stored as 
```python
MultiIndex(levels=[[1], [1], [3]],
           labels=[[0, 0], [0, 0], [0, -1]],
           names=[u'A', u'B', u'C'])
```
with a `NaN` placeholder of -1.

These label values are what get passed to the sorting algorithm for both [DataFrames](https://github.com/pandas-dev/pandas/blob/master/pandas/core/frame.py#L3334-L3337) and [Series](https://github.com/pandas-dev/pandas/blob/master/pandas/core/series.py#L1782).  Since the sorting only happens on the `labels`, it has no notion of the `NaN`.

This has been discussed in #14015 and #14672 .

My original naive solution was to change [these lines](https://github.com/pandas-dev/pandas/blob/master/pandas/core/frame.py#L3337-L3338) from:

```python
indexer = _lexsort_indexer(labels.labels, orders=ascending,
                                            na_position=na_position)
```

to

```python
index_values_list = np.dstack(labels.get_values())[0].tolist()
indexer = _lexsort_indexer(index_values_list, orders=ascending,
                                           na_position=na_position)
```

This didn't break any tests, but it isn't necessarily the best approach.  

#### Expected Output

```python
In [7]: df.sort_index(na_position="last")
Out[7]: 
         0  1
A B C        
1 1 3    1  2
    NaN  3  4

```

#### Output of ``pd.show_versions()``

<details>
In [3]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Linux
OS-release: 3.16.0-77-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.19.0
nose: 1.3.4
pip: 9.0.0
setuptools: 27.2.0
Cython: 0.21
numpy: 1.11.2
scipy: 0.16.1
statsmodels: 0.6.1
xarray: None
IPython: 4.0.0
sphinx: 1.2.3
patsy: 0.3.0
dateutil: 2.5.3
pytz: 2016.7
blosc: None
bottleneck: None
tables: 3.1.1
numexpr: 2.3.1
matplotlib: 1.5.0
openpyxl: 1.8.5
xlrd: 0.9.3
xlwt: 0.7.5
xlsxwriter: 0.5.7
lxml: 3.4.0
bs4: 4.3.2
html5lib: None
httplib2: 0.9.2
apiclient: 1.5.5
sqlalchemy: 0.9.7
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
jinja2: 2.7.3
boto: 2.32.1
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by linebp**
- [X] closes #14784 
 - [X] tests added / passed
 - [X] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
@linebp I'd prefer if you did something like I illustrated here: https://github.com/pandas-dev/pandas/issues/14784#issuecomment-287362444

Put these tests (and on Series as well) in ``test_multilevel.py`` where these types of things live. Further Ideally we'd do a lower-level test in ``pandas/tests/indexes/tests/test_multi.py`` and/or ``pandas/tests/tests_sorting.py``

**Comment by linebp** [comment]
I am not sure what I am missing, I tried to do as suggested. These are my steps figuring out what to do:
* I looked for a function called `_lexsort_indexer` and didn't manage to locate it. 
* I figured what you meant was the function `lexsort_indexer`, which made sense to me, as this function is called while sorting the index.
* The `lexsort_indexer` is called directly from `sort_index`, no code in MultiIndex is called(?)
* I put the fix in `sort_index` as that is the function calling `lexsort_indexer`. This made good sense to me, as `lexsort_indexer` is not actually broken, but the data passed to it is. 

Where and how did I go wrong?

I'll move the tests to a more appropriate location! ... and test the Series sorting too.

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3392,7 +3392,13 @@ def sort_index(self, axis=0, level=None, ascending=True, inplace=False,
             if not labels.is_lexsorted():
                 labels = MultiIndex.from_tuples(labels.values)
 
-            indexer = lexsort_indexer(labels.labels, orders=ascending,
```
Comment (jreback) [inline pandas/core/frame.py]: as I said, you can simply create a method on ``MultiIndex``, called ``_lexsort_indexer`` (let's make it private), that returns the correct indexer.

**Review by jreback — CHANGES_REQUESTED** [review]
this needs testing for Series (which won't work as you have changed the code)

File: pandas/core/frame.py [hunk 2]
```diff
@@ -3385,15 +3385,9 @@ def sort_index(self, axis=0, level=None, ascending=True, inplace=False,
                                                  sort_remaining=sort_remaining)
 
         elif isinstance(labels, MultiIndex):
-            from pandas.core.sorting import lexsort_indexer
-
-            # make sure that the axis is lexsorted to start
-            # if not we need to reconstruct to get the correct indexer
```
Comment (jreback) [inline pandas/core/frame.py]: so leave the ``is_lexsorted`` check in here; this is fixed in #15694 w/o this subtle things will break.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -3385,15 +3385,9 @@ def sort_index(self, axis=0, level=None, ascending=True, inplace=False,
...
-            if not labels.is_lexsorted():
-                labels = MultiIndex.from_tuples(labels.values)
 
-            indexer = lexsort_indexer(labels.labels, orders=ascending,
-                                      na_position=na_position)
+            indexer = labels._lexsort_indexer(orders=ascending,
```
Comment (jreback) [inline pandas/core/frame.py]: just do

```
indexer = lexsort_indexer(labels._get_labels_for_sorting(), orders=ascending,
                                      na_position=na_position)
```

so IOW rename ``pandas.indexes.multi._lexsort_indexer`` (your new routines) to ``_get_labels_for_sorting``

File: pandas/core/series.py [hunk 4]
```diff
@@ -1763,8 +1763,8 @@ def sort_index(self, axis=0, level=None, ascending=True, inplace=False,
             new_index, indexer = index.sortlevel(level, ascending=ascending,
                                                  sort_remaining=sort_remaining)
         elif isinstance(index, MultiIndex):
-            from pandas.core.sorting import lexsort_indexer
```
Comment (jreback) [inline pandas/core/series.py]: make the same change as above

File: pandas/indexes/multi.py [hunk 5]
```diff
@@ -1499,6 +1499,24 @@ def reorder_levels(self, order):
...
+        from pandas.core.sorting import lexsort_indexer
+        from pandas.core.categorical import Categorical
+        from pandas.core.algorithms import unique
+
+        # if not self.is_lexsorted():
+        #     mi = MultiIndex.from_tuples(self.values)
+        # else:
+        #     mi = self
+
+        keys = []
```
Comment (jreback) [inline pandas/indexes/multi.py]: rename to ``_get_labels_for_sorting()``

add a doc-string.

rename ``keys`` -> ``labels``

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by linebp** [comment]
so, add the `is_lexsorted` check to both series and frames `sort_index` right before call to `_lexsort_indexer`? What is the reason for not adding the check to the `_lexsort_indexer` function?

Panels seems to have a function similar to `sort_index` on frames and series, is that in the scope of this bug?

If you use the `level` option when calling `sort_index`, the `na_position` option is ignored.  Should this be fixed to?

**Comment by jreback** [comment]
@linebp this needs to be very targeted. don't touch Panel and don't worry about level.

instead of a``labels.labels``, call ``labels._get_labels_for_sorting()``

it is *very* easy to break things w/o very specific patches.

**Comment by jreback** [comment]
@linebp maybe I wasn't clear. These ALL the existing code, except for

``indexer = lexsort_indexer(index._get_labels_for_sorting(), orders=ascending)``

where ``_get_labels_for_sorting()`` is the routine that you added ``_lexsort_indexer``

so you are adding 1 function to MultiIndex, and changing 1 line in series & 1 line in frame.py thats it.

**Comment by linebp** [comment]
I am still a bit confused about what to do, but this adds 1 function and only changes 1 line in series and frame each. No comments on the tests, so I let those stay as is.

**Comment by jreback** [comment]
ok, pls rebase this as I pushed in a bunch of fixes for sorting (generally)

**Comment by linebp** [comment]
I made an effort to rebase, how is it looking now?

**Comment by jreback** [comment]
@linebp code looks good. pls add a whatsnew (bug fix). make sure you rebase on origin (there are still conflicts showing)

```
git rebase -i thisbranch origin/master
```

then push with ``-f`` to update.

**Review by jreback — APPROVED** [review]

**Comment by linebp** [comment]
It fails when `Categorical.from_codes()` is called on an index where some rows are removed, for example with `dropna()`. The index is not relabelled and so the labels may be outside the -1 to len(categories) range. Example:
 
```
import pandas as pd
import numpy as np

df = pd.DataFrame({
    'date': pd.to_datetime([
        '20121002', '20121007', '20130130', '20130202', '20130305',
        '20121002', '20121207', '20130130', '20130202', '20130305',
        '20130202', '20130305'
    ]),
    'user_id': [1, 1, 1, 1, 1, 3, 3, 3, 5, 5, 5, 5],
    'whole_cost': [1790, np.nan, 280, 259, np.nan, 623, 90, 312, np.nan, 301,
                   359, 801],
    'cost': [12, 15, 10, 24, 39, 1, 0, np.nan, 45, 34, 1, 12]
}).set_index(['date', 'user_id'])

df.dropna().sort_index()

```

**Comment by jreback** [comment]
@linebp I pushed a fix to your branch. can you add the above as an additional tests. ping on green.

**Pull request closed without merging**